File: automaguard-java/src/main/java/io/automaguard/EnforcementException.java
package io.automaguard;

/**
 * Thrown when policy evaluation returns a {@link Verdict#DENY} verdict.
 *
 * <p>The full {@link PolicyResult} is attached so callers can inspect
 * triggered rules, invariant violations, and the denial reason without
 * consulting a separate log stream.</p>
 *
 * <pre>{@code
 * try {
 *     String result = client.prompt(prompt).call().content();
 * } catch (EnforcementException ex) {
 *     System.err.println("Blocked: " + ex.getMessage());
 *     ex.getResult().getViolations().forEach(v ->
 *         System.err.println("  Invariant: " + v));
 * }
 * }</pre>
 */
public final class EnforcementException extends RuntimeException {

    private final PolicyResult result;

    /**
     * Construct from a {@link PolicyResult} with a {@link Verdict#DENY} verdict.
     * The exception message is the denial reason, or a generic fallback if none
     * was provided.
     */
    public EnforcementException(PolicyResult result) {
        super(result.getReason() != null
                ? result.getReason()
                : "Policy denied: " + result.getVerdict());
        this.result = result;
    }

    /**
     * Construct with a custom message alongside the full result.
     */
    public EnforcementException(String message, PolicyResult result) {
        super(message);
        this.result = result;
    }

    /** The full evaluation result that triggered this exception. */
    public PolicyResult getResult() { return result; }
}

File: automaguard-dotnet/src/AutomaGuard/EnforcementException.cs
namespace AutomaGuard;

/// <summary>
/// Thrown when a policy evaluation returns a <see cref="Verdict.Deny"/> verdict.
///
/// The full <see cref="PolicyResult"/> is attached so callers can inspect
/// triggered rules, violations, and the denial reason.
/// </summary>
public sealed class EnforcementException : Exception
{
    /// <summary>The full policy evaluation result that triggered this exception.</summary>
    public PolicyResult Result { get; }

    /// <summary>
    /// Initialise the exception from a <see cref="PolicyResult"/> with a
    /// <see cref="Verdict.Deny"/> verdict.
    /// </summary>
    /// <param name="result">The denied evaluation result.</param>
    public EnforcementException(PolicyResult result)
        : base(result.Reason ?? $"Policy denied: {result.Verdict}")
    {
        Result = result;
    }

    /// <summary>
    /// Initialise the exception with a custom message and a full result.
    /// </summary>
    public EnforcementException(string message, PolicyResult result)
        : base(message)
    {
        Result = result;
    }
}
